core/lib/vlog/src/lib.rs

```diff
@@ -92,28 +171,108 @@ impl ObservabilityBuilder {
+
+        let tracer = opentelemetry_otlp::new_pipeline()
+            .tracing()
+            .with_exporter(
+                opentelemetry_otlp::new_exporter()
+                    .http()
+                    .with_endpoint(otlp_endpoint),
+            )
+            .with_trace_config(
+                trace::config()
+                    .with_sampler(Sampler::AlwaysOn)
+                    .with_id_generator(RandomIdGenerator::default())
+                    .with_resource(Resource::new(resource)),
+            )
+            .install_batch(opentelemetry::runtime::Tokio)
+            .unwrap();
+
+        opentelemetry::global::set_text_map_propagator(TraceContextPropagator::new());
+        let layer = tracing_opentelemetry::layer()
+            .with_tracer(tracer)
+            .with_filter(filter);
+        subscriber.with(layer)
+    }
+
     /// Initializes the observability subsystem.
     pub fn build(self) -> ObservabilityGuard {
         // Initialize logs.
         match self.log_format {
             LogFormat::Plain => {
-                tracing_subscriber::registry()
+                let subscriber = tracing_subscriber::registry()
                     .with(tracing_subscriber::EnvFilter::from_default_env())
-                    .with(fmt::Layer::default())
-                    .init();
+                    .with(fmt::Layer::default());
+                if let Some(opts) = self.opentelemetry_options {
+                    let subscriber = Self::add_opentelemetry_layer(
+                        opts.opentelemetry_level,
+                        opts.otlp_endpoint,
+                        opts.service_name,
+                        subscriber,
+                    );
+                    subscriber.init()
+                } else {
+                    subscriber.init()
+                }
             }
             LogFormat::Json => {
                 let timer = tracing_subscriber::fmt::time::UtcTime::rfc_3339();
-                tracing_subscriber::registry()
+                let subscriber = tracing_subscriber::registry()
                     .with(tracing_subscriber::EnvFilter::from_default_env())
                     .with(
                         fmt::Layer::default()
                             .with_file(true)
                             .with_line_number(true)
                             .with_timer(timer)
                             .json(),
-                    )
-                    .init();
+                    );
+                if let Some(opts) = self.opentelemetry_options {
+                    let subscriber = Self::add_opentelemetry_layer(
+                        opts.opentelemetry_level,
+                        opts.otlp_endpoint,
+                        opts.service_name,
+                        subscriber,
+                    );
+                    subscriber.init()
+                } else {
+                    subscriber.init()
+                }
```

Comment:
This seems to be duplicated. I think it can be avoided if `subscriber` is moved outside of the match statement.

Reply:
It's actually not that simple as the two subscribers have slightly different types (`tracing` stores information about the layers inside the generic). Same reason why `susbcriber.init()` here can't be brought outside of `if`. I can erase the type by boxing it, but not sure if it's worth the trouble. WDYT?

Reply:
No strong opinion.